types/react/index.d.ts

```diff
@@ -74,7 +74,9 @@ declare namespace React {
 
     type JSXElementConstructor<P> =
         | ((props: P) => ReactElement | null)
-        | (new (props: P) => Component<P, any>);
+        | (new (props: P) => Component<P, any>)
+        | DRFC<P>
+        | (new (props: P) => DRPureComponent<P, any>);
```

Comment:
```suggestion
        | (new (props: DeepReadonly<P>) => DRPureComponent<P, any>);
```

Why does this only produce pure components? What about regular components?

Reply:
DRPureComponent wraps P in DeepReadonly. I guess the benefit of your suggestions is that it prevents the constructor function itself from mutating P - right? 
As for only adding DRPureComponent, I'm not sure about it any more. As I wrote yesterday (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36066) immutable data goes together with `PureComponent`. If the data is immutable we can trust that if the references in props hasn't changed, then the object hasn't changed, there is no reason to rerender the component, and it makes sense to use a PureComponent instead. 
However, the fact that our component won't mutate Props doesn't mean that no one will mutate Props, as it could be mutated before it reaches our component. The new component types make it more compelling to use a DRPureComponent, but you are right there probably is a need for a new DRComponent that just extends the "regular" Component.

Reply:
Yeah I was mostly just wondering what the scope of this change would be. Theoretically no props should be mutated so we could change nearly everything to be deep-readonly, so I was just trying to get a sense of that scope.

Reply:
Following this discussion I realized that Props isn't protected as DeepReadonly in the constructor, and could be mutated there. Can we remedy this?

Reply:
It *should* be okay if you change the constructor of the `Component` class to have a deep-readonly props, but I'm not 100% sure as deep-readonly types are subtypes of non-readonly, and some weird assignability checks might fail unrelatedly.

Reply:
Following this discussion I added a new type DRComponent. 
As for your suggestion, since every type should extend it's equivalent DeepReadonly type, your suggestion seems valid.